src/cargo/core/resolver/mod.rs

```diff
@@ -712,47 +704,38 @@ struct BacktrackFrame {
-/// is defined within a `Context`.
+/// is defined within a `Context`. In addition, the iterator will return
+/// candidates that already have been activated first.
 ///
 /// Candidates passed to `new` may not be returned from `next` as they could be
-/// filtered out, and as they are filtered the causes will be added to `conflicting_prev_active`.
+/// filtered out. The filtered candidates and the causes will be returned by `new`.
 #[derive(Clone)]
 struct RemainingCandidates {
-    remaining: RcVecIter<Summary>,
+    prioritized_candidates: RcVecIter<Summary>,
     // This is an inlined peekable generator
     has_another: Option<Summary>,
 }
 
 impl RemainingCandidates {
-    fn new(candidates: &Rc<Vec<Summary>>) -> RemainingCandidates {
-        RemainingCandidates {
-            remaining: RcVecIter::new(Rc::clone(candidates)),
-            has_another: None,
-        }
-    }
-
-    /// Attempts to find another candidate to check from this list.
-    ///
-    /// This method will attempt to move this iterator forward, returning a
-    /// candidate that's possible to activate. The `cx` argument is the current
-    /// context which determines validity for candidates returned, and the `dep`
-    /// is the dependency listing that we're activating for.
-    ///
-    /// If successful a `(Candidate, bool)` pair will be returned. The
-    /// `Candidate` is the candidate to attempt to activate, and the `bool` is
-    /// an indicator of whether there are remaining candidates to try of if
-    /// we've reached the end of iteration.
-    ///
-    /// If we've reached the end of the iterator here then `Err` will be
-    /// returned. The error will contain a map of package ID to conflict reason,
-    /// where each package ID caused a candidate to be filtered out from the
-    /// original list for the reason listed.
-    fn next(
-        &mut self,
-        conflicting_prev_active: &mut ConflictMap,
+    /// Prefilters and sorts the list of candidates to determine which ones are
+    /// valid to activate, and which ones should be prioritized.
+    fn new(
+        candidates: &Rc<Vec<Summary>>,
         cx: &ResolverContext,
-    ) -> Option<(Summary, bool)> {
-        for b in self.remaining.by_ref() {
+    ) -> (RemainingCandidates, ConflictMap) {
+        // `conflicting_activations` stores all the reasons we were unable to
+        // activate candidates. One of these reasons will have to go away for
+        // backtracking to find a place to restart. It is also the list of
+        // things to explain in the error message if we fail to resolve.
+        //
+        // This is a map of package ID to a reason why that packaged caused a
+        // conflict for us.
+        let mut conflicting_activations = ConflictMap::new();
+
+        let mut activated_candidates: Vec<Summary> = Vec::with_capacity(candidates.len());
+        let mut non_activated_candidates: Vec<Summary> = Vec::with_capacity(candidates.len());
```

Comment:
I'm a little confused why we need two vectors here (that get concatenated) instead of just sorting one vec. and if it's just a sort, can it be combined with the existing sort at https://github.com/rust-lang/cargo/blob/ecd68bba3491ecc374d30bf894d8bc85f4119bb6/src/cargo/core/resolver/version_prefs.rs#L67

Reply:
It's not quite a sort. We're actually just splitting the vector into two buckets, while preserving the order of the elements. 

While this can technically be done in place, you'd still need a tmp array to copy out segments. Here is the performance comparison:

N = activations
M = total candidates

### Current implementation with split arrays

item memcopies = 2 * M
array mallocs = 3

  
### In-place implementation

item memcopies = N * (M + 2)
array mallocs = 1

Reply:
Yes but:
- `candidates.sort_by_key(|c| cx.activated.contains(c))` is a lot shorter to read.
- It takes a lot of pointer (`Summary`) copying to compete with the cost of two allocations.
- We already have code `sort_summaries` sorting this array.

Overall we should merge whichever solution is the easiest to understand.

Reply:
`sort_by_key` doesn't do what we need.

For example, when I run it on that unit test I just wrote above it will yield:
```
assertion `left == right` failed
  left: [1, 3, 5, 7, 9, 2, 4, 6, 8]
 right: [2, 4, 6, 8, 1, 3, 5, 7, 9]
 ```
 
We need to preserve order of the two array segments. I don't see a way to do this without at least 1 allocation.

Reply:
Ok that's neat. I didn't realize `sort_by_key` would maintain segment order when used with a boolean predicate.

Reply:
All the sort functions maintain insertion order anytime the predicates are equal. They each come with a `_unstable` variant that is a little faster but does not uphold this property. (And yes, it's unfortunate that `unstable` has a specific meaning in rust and a different specific meaning in sorting algorithms.)

Reply:
Ok after poking around a bit I realized we cannot filter & sort in-place. The outer loop relies on the original `candidates` vector to generate the Error here:

```rust
debug!("no candidates found");
Err(errors::activation_error(
    &resolver_ctx,
    registry.registry,
    &parent,
    &dep,
    &conflicting_activations,
    &candidates,
    gctx,
))
```

That means if we're pre-filtering the array, we must alloc a new one and copy the pointers over. 

On the topic of using `sort_by_key()`, after playing around with it I actually think the way it's currently done is better because we are able to both sort and filter based on the result of one comparison. If we did it as `filter().sort_by_key()` we'd be doing an unnecessary extra comparison on each item instead of relying on the else condition.

As to how to remove the extra allocation I'm currently doing; I could make the container be something like a VecDeque, where the "prioritized" candidates are pushed to the back and the "non-prioritized" candidates are pushed to the front, then use a zipped iterator that iterates in reverse until it reaches the "partition point", after which it will iterate over the front elements normally.

Reply:
The resolver is pretty fast these days. On one hand that took a lot of focus on micro-optimizations in these hot loops. On the other hand that means we can afford to write things "slower but more readable". Let's figure out what we want before geeking out too much about how to get all the performance out of it?